applications/FluidDynamicsApplication/custom_processes/weighted_divergence_calculation_process.cpp

```diff
@@ -0,0 +1,215 @@
+                    GeometryUtils::ShapeFunctionsGradients(rDN_De, InvJ0, DN_DX);
+
+                    // Compute local gradient
+                    const Vector grad_x = prod(trans(DN_DX), values_x);
+                    const Vector grad_y = prod(trans(DN_DX), values_y);
+                    const Vector grad_z = prod(trans(DN_DX), values_z);
+
+                    // Compute divergence and velocity seminorm
+                    const double aux_current_divergence = ComputeAuxiliaryElementDivergence(grad_x, grad_y, grad_z);
+                    const double aux_current_velocity_seminorm = ComputeAuxiliaryElementVelocitySeminorm(grad_x, grad_y, grad_z);
+                    const double gauss_point_volume = r_integration_points[point_number].Weight() * detJ0;
+                    divergence_current += std::pow(aux_current_divergence,2) * gauss_point_volume;
+                    velocity_seminorm_current += aux_current_velocity_seminorm * gauss_point_volume;
+                }
+
+                // Retrieve divergence from previous time step
+                const double divergence_old = it_elem->GetValue(DIVERGENCE);
+                const double velocity_seminorm_old = it_elem->GetValue(VELOCITY_H1_SEMINORM);
+
+                // Compute divergence weighted time average
+                auto divergence_current_avg = ComputeWeightedTimeAverage(divergence_old, divergence_current);
+                it_elem->SetValue(DIVERGENCE,divergence_current_avg);
+
+                // Compute divergence_norm weighted time average
+                auto velocity_seminorm_current_avg = ComputeWeightedTimeAverage(velocity_seminorm_old, velocity_seminorm_current);
+                it_elem->SetValue(VELOCITY_H1_SEMINORM,velocity_seminorm_current_avg);
+            }
+        }
+
+        KRATOS_CATCH("");
+    }
+
+    // Compute time average
+    double WeightedDivergenceCalculationProcess::ComputeWeightedTimeAverage(const double old_average, const double current_value)
+    {
+        // Extract time information
+        const auto& r_current_process_info = mrModelPart.GetProcessInfo();
+        const double& time_step_current  = r_current_process_info[TIME];
+        const auto& r_previous_process_info = r_current_process_info.GetPreviousTimeStepInfo();
+        const double& time_step_previous = r_previous_process_info[TIME];
+        const double& final_time = r_current_process_info[END_TIME];
+
+        const double new_average = std::sqrt(((time_step_previous-mTimeCoefficient*final_time) * std::pow(old_average,2) + (time_step_current - time_step_previous) * current_value) / (time_step_current - mTimeCoefficient*final_time));
+        return new_average;
+    }
+
+    // Compute local auxiliar divergence
+    double WeightedDivergenceCalculationProcess::ComputeAuxiliaryElementDivergence(const Vector grad_x, const Vector grad_y, const Vector grad_z)
+    {
+        const double aux_current_divergence = grad_x[0] + grad_y[1] + grad_z[2];
```

Comment:
Here we are accessing the third position of grad_z ([2]), but in 2D this grad_z is a 2x1 vector (in FullDebug this throws an error but not in Release). Can you check that the sizes are correct for 2D and 3D?

Reply:
Done

Reply:
In the change I did using ShapeFunctionsIntegrationPointsGradients, the resize is done now automatically in the ShapeFunctionsIntegrationPointsGradients method